packages/flutter_tools/lib/src/flutter_plugins.dart

```diff
@@ -1137,127 +1137,176 @@ List<PluginInterfaceResolution> resolvePlatformImplementation(
-      final Iterable<Plugin> directDependencies = candidates.where((Plugin plugin) {
-        return plugin.isDirectDependency;
-      });
-      if (directDependencies.isNotEmpty) {
-        if (directDependencies.length > 1) {
-          globals.printError(
-              'Plugin ${entry.key}:$platformKey has conflicting direct dependency implementations:\n'
-                  '${directDependencies.map((Plugin plugin) => '  ${plugin.name}\n').join()}'
-                  'To fix this issue, remove all but one of these dependencies from pubspec.yaml.\n'
-          );
-          hasResolutionError = true;
-        } else {
-          pluginResolution.add(directDependencies.first);
+    // Now resolve all the possible resolutions to a single option for each
+    // plugin, or throw if that's not possible.
+    for (final MapEntry<String, List<Plugin>> implCandidatesEntry in pluginImplCandidates.entries) {
+      try {
+        final Plugin? resolution = _resolveImplementationOfPlugin(
+          platformKey,
+          implCandidatesEntry,
+          defaultPackageName: defaultImplementations[implCandidatesEntry.key],
+        );
+        if (resolution != null) {
+          pluginResolution[implCandidatesEntry.key] = resolution;
         }
-        continue;
-      }
-      // Next, defer to the default implementation if there is one.
-      final String? defaultPackageName = defaultImplementations[entry.key];
-      if (defaultPackageName != null) {
-        final int defaultIndex = candidates
-            .indexWhere((Plugin plugin) => plugin.name == defaultPackageName);
-        if (defaultIndex != -1) {
-          pluginResolution.add(candidates[defaultIndex]);
-          continue;
+      } on ToolExit catch (e) {
+        if (e.message != null) {
+          globals.printError(e.message!);
         }
-      }
-      // Otherwise, require an explicit choice.
-      if (candidates.length > 1) {
-        globals.printError(
-            'Plugin ${entry.key}:$platformKey has multiple possible implementations:\n'
-                '${candidates.map((Plugin plugin) => '  ${plugin.name}\n').join()}'
-                'To fix this issue, add one of these dependencies to pubspec.yaml.\n'
-        );
         hasResolutionError = true;
-        continue;
       }
     }
 
-    finalResolution.addAll(
-      pluginResolution.map((Plugin plugin) =>
+    pluginResolutions.addAll(
+      pluginResolution.values.map((Plugin plugin) =>
           PluginInterfaceResolution(plugin: plugin, platform: platformKey)),
     );
   }
   if (hasResolutionError) {
     throwToolExit('Please resolve the plugin implementation selection errors');
   }
-  return finalResolution;
+  return pluginResolutions;
+}
+
+/// Determine whether this plugin serves as implementation for an app-facing
+/// package with [implementsPluginName] or/and whether it references a default
+/// plugin [defaultImplPluginName] for the given platform [platformKey].
```

Comment:
I'm still confused as to why this method does two basically unrelated things. Whether something implements another plugin and whether it defines default implementations for a plugin are almost entirely orthogonal. The only intersection point is inline implementations, but it seems like that could be much more cleanly handled by extracting the conditional checks for whether a plugin is self-implementing to a helper, and calling that from two different methods, each one of which only does one thing.

Reply:
> but it seems like that could be much more cleanly handled by extracting the conditional checks

> and calling that from two different methods, each one of which only does one thing

I totally agree! That was also my goal. But I'm conflicting in two directions. If changing too much of the code, noone can follow the refactor steps, I have taken to get there. Saying now, this can be changed is more trivial, because we are right in the middle of the process. But untangling all the logic in a single step, didn't  seem helpful to me, to keep the refactoring steps small and simple. But sure, I'm happy to do that.

Reply:
I splitted that in 1c0aff6042587695feae7d87539f05e31fbcb56a now. But the logic is not exactly as before. I introduced constraints to draw attention to these cases in fbaa6f32bb3fac20e8cfd9d9feeab93ab7646fe2 and c316a603200a7a9d14ea4868a265039e517d4a89